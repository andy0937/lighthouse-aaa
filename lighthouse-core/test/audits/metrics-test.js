/**
 * @license Copyright 2018 Google Inc. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */
'use strict';

const MetricsAudit = require('../../audits/metrics.js');
const TTIComputed = require('../../computed/metrics/interactive.js');

const pwaTrace = require('../fixtures/traces/progressive-app-m60.json');
const pwaDevtoolsLog = require('../fixtures/traces/progressive-app-m60.devtools.log.json');

/* eslint-env jest */

describe('Performance: metrics', () => {
  it('evaluates valid input correctly', async () => {
    const artifacts = {
      traces: {
        [MetricsAudit.DEFAULT_PASS]: pwaTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: pwaDevtoolsLog,
      },
    };

    const context = {settings: {throttlingMethod: 'simulate'}, computedCache: new Map()};
    const result = await MetricsAudit.audit(artifacts, context);
    expect(result.details.items[0]).toMatchSnapshot();
  });

  it('does to fail the entire audit when TTI errors', async () => {
    const artifacts = {
      traces: {
        [MetricsAudit.DEFAULT_PASS]: pwaTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: pwaDevtoolsLog,
      },
    };

    const mockTTIFn = jest.spyOn(TTIComputed, 'request');
    mockTTIFn.mockRejectedValueOnce(new Error('TTI failed'));
    const context = {settings: {throttlingMethod: 'simulate'}, computedCache: new Map()};
    const result = await MetricsAudit.audit(artifacts, context);
    expect(result.details.items[0].interactive).toEqual(undefined);
  });
});
