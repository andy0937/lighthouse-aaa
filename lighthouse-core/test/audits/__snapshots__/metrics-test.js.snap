// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Performance: metrics evaluates valid input correctly 1`] = `
Object {
  "estimatedInputLatency": 78,
  "estimatedInputLatencyTs": undefined,
  "firstCPUIdle": 3351,
  "firstCPUIdleTs": undefined,
  "firstContentfulPaint": 1307,
  "firstContentfulPaintTs": undefined,
  "firstMeaningfulPaint": 1511,
  "firstMeaningfulPaintTs": undefined,
  "interactive": 3427,
  "interactiveTs": undefined,
  "observedDomContentLoaded": 560,
  "observedDomContentLoadedTs": 225414732309,
  "observedFirstContentfulPaint": 499,
  "observedFirstContentfulPaintTs": 225414670885,
  "observedFirstMeaningfulPaint": 783,
  "observedFirstMeaningfulPaintTs": 225414955343,
  "observedFirstPaint": 499,
  "observedFirstPaintTs": 225414670868,
  "observedFirstVisualChange": 520,
  "observedFirstVisualChangeTs": 225414692015,
  "observedLastVisualChange": 818,
  "observedLastVisualChangeTs": 225414990015,
  "observedLoad": 2199,
  "observedLoadTs": 225416370913,
  "observedNavigationStart": 0,
  "observedNavigationStartTs": 225414172015,
  "observedSpeedIndex": 605,
  "observedSpeedIndexTs": 225414776724,
  "observedTraceEnd": 12540,
  "observedTraceEndTs": 225426711887,
  "speedIndex": 1657,
  "speedIndexTs": undefined,
  "totalBlockingTime": 726,
}
`;
